### DISCUSSION TITLE
Throwing error when using Map

### DISCUSSION THREAD

**Discussion by fregin** [ExpediaGroup/graphql-kotlin#931] [Q&A] [answered]
**Library Version**
3.6.6

**Describe the bug**
I created the graphql fun with `Map<String, Any>` field.
And it throws error: 
```
Caused by: java.lang.ClassCastException: class sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl cannot be cast to class java.lang.Class (sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl and java.lang.Class are in module java.base of loader 'bootstrap')
	at com.expediagroup.graphql.generator.extensions.KTypeExtensionsKt.getJavaClass(kTypeExtensions.kt:39) ~[graphql-kotlin-schema-generator-3.6.1.jar:3.6.1]
	at com.expediagroup.graphql.generator.extensions.KParameterExtensionsKt.isListType(kParameterExtensions.kt:29) ~[graphql-kotlin-schema-generator-3.6.1.jar:3.6.1]
```
I debuged this at got totally sure that it happens because of map field.

**To Reproduce**
I have this problem on my spring application.
```
@Component
class MyMutation : Mutation {
    suspend fun myFun(vars: Map<String, Any>? = null) = vars.toString()
}
```

**Expected behavior**
Not throwing such error. I think it's not normal to send error while you are checking that the field is array or not.

**How to solve**
I think I have a workaround. Just improve KTypeExtensions.kt file.

In function `getJavaClass` there should be a body:
```
internal fun KType.getJavaClass() = 
    if (javaType::class.isSubclassOf(ParameterizedType::class)) (classifier as KClass<*>).java
    else javaType as Class<*>
```

I can try to fix it, if you will allow me

**Comment by dariuszkuc** [comment, marked as answer]
One of the core principles of GraphQL is the strong type safety. When you use `Map` for storing arbitrary objects it breaks that core assumption - e.g. how will your clients know what are the types stored in the map or in your case how does the client knows what are accepted input parameters? In general, instead of arbitrary map of objects you should be able to model your schema to have more meaningful type.

Currently GraphQL spec does not support arbitrary `Map` type (see https://github.com/graphql/graphql-spec/issues/101 for details) so unfortunately we won't be adding support for it to this library.

---
NOTE: You should be able to add support to a generic `Map` type to your schema using schema generator hook.

**Reply by fregin** [reply]
Ok, if we are talking about not map, any other generic type.
I don't think it's good when the code is throwing exceptions when it checks for array types, even if it cannot support generics at all.
It must throwing other error, that describes the real problem

P.S. I used custom scalar for map, but it's ok for output, and doesn't work for input

**Reply by dariuszkuc** [reply]
Generics are not supported by the spec either (see https://github.com/ExpediaGroup/graphql-kotlin/issues/292 and linked https://github.com/graphql/graphql-spec/issues/190). 

I agree that exceptions should be meaningful - if you see something that could be improved feel free to submit issue/pr.

---
Custom scalar map for input should work but you would also probably need to provide custom data fetcher/resolver to correctly deserialize the input (in addition to the hook logic).

**Reply by fregin** [reply]
Ok, got you
I will try to find some time to make a PR

Thanks
